aptos-move/framework/aptos-stdlib/sources/data_structures/smart_table.move

```diff
@@ -541,4 +623,95 @@ module aptos_std::smart_table {
+        let starting_vector_index = 0;
+        let (keys, starting_bucket_index_r, starting_vector_index_r) = keys_paginated(
+            &table,
+            starting_bucket_index,
+            starting_vector_index,
+            0
+        );
+        assert!(starting_bucket_index_r == option::none(), 0);
+        assert!(starting_vector_index_r == option::none(), 0);
+        assert!(vector::is_empty(&keys), 0);
+        while (i < 100) {
+            add(&mut table, i, 0);
+            vector::push_back(&mut expected_keys, i);
+            i = i + 1;
+        };
+        let keys = keys(&table);
+        assert!(vector::length(&keys) == vector::length(&expected_keys), 0);
+        vector::for_each_ref(&keys, |e_ref| {
+            assert!(vector::contains(&expected_keys, e_ref), 0);
+        });
+        let keys = vector[];
+        let starting_bucket_index = 0;
+        let starting_vector_index = 0;
+        let returned_keys = vector[];
+        vector::length(&returned_keys); // To eliminate erroneous compiler "unused" warning
+        loop {
+            (returned_keys, starting_bucket_index_r, starting_vector_index_r) = keys_paginated(
+                &table,
+                starting_bucket_index,
+                starting_vector_index,
+                15
+            );
+            vector::append(&mut keys, returned_keys);
+            if (
+                starting_bucket_index_r == option::none() ||
+                starting_vector_index_r == option::none()
+            ) break;
+            starting_bucket_index = option::destroy_some(starting_bucket_index_r);
+            starting_vector_index = option::destroy_some(starting_vector_index_r);
+        };
+        assert!(vector::length(&keys) == vector::length(&expected_keys), 0);
+        vector::for_each_ref(&keys, |e_ref| {
+            assert!(vector::contains(&expected_keys, e_ref), 0);
+        });
+        destroy(table);
+        table = new();
+        add(&mut table, 1, 0);
+        add(&mut table, 2, 0);
+        (keys, starting_bucket_index_r, starting_vector_index_r) = keys_paginated(&table, 0, 0, 1);
+        (returned_keys, starting_bucket_index_r, starting_vector_index_r) = keys_paginated(
+            &table,
+            option::destroy_some(starting_bucket_index_r),
+            option::destroy_some(starting_vector_index_r),
+            1,
+        );
+        vector::append(&mut keys, returned_keys);
+        assert!(keys == vector[1, 2] || keys == vector[2, 1], 0);
+        assert!(starting_bucket_index_r == option::none(), 0);
+        assert!(starting_vector_index_r == option::none(), 0);
+        (keys, starting_bucket_index_r, starting_vector_index_r) = keys_paginated(&table, 0, 0, 0);
+        assert!(keys == vector[], 0);
+        assert!(starting_bucket_index_r == option::none(), 0);
+        assert!(starting_vector_index_r == option::none(), 0);
+        destroy(table);
+    }
+
+    #[test, expected_failure(abort_code = EINVALID_BUCKET_INDEX)]
```

Comment:
could you try more corner cases with index 0 or keys num 0?

Reply:
Addressed in 809e8d2b36